### PR TITLE
Allow issuing model commands via PASE

### PR DESCRIPTION
**Pull request by markus-becker-tridonic-com**
#### Problem

chip-tool allows to only create a PASE session (e.g. via BLE).

chip-tool does now allow to issue commands via PASE, although the spec allows it and chip-tool auto-commissioner is using it internally.

#### Change overview

* Add an option `--via-pase 1` which allows to send commands via PASE.
* Documentation for PASE only.
* changes protection level of FindCommissioneeDevice (needs to be discussed)

#### Testing

How was this tested? (at least one bullet point required)
* Manually on POSIX against nRF, as described in the README.md

### CODE REVIEW

File: examples/chip-tool/commands/clusters/ModelCommand.cpp [hunk 1]
```diff
@@ -35,11 +35,20 @@ CHIP_ERROR ModelCommand::RunCommand()
         return SendGroupCommand(GroupIdFromNodeId(mNodeId), fabricIndex);
     }
 
-    ChipLogProgress(chipTool, "Sending command to node 0x%" PRIx64, mNodeId);
-    return CurrentCommissioner().GetConnectedDevice(mNodeId, &mOnDeviceConnectedCallback, &mOnDeviceConnectionFailureCallback);
+    if (mViaPase.ValueOr(0) == 1)
+    {
+        ChipLogProgress(chipTool, "Sending command via PASE to node 0x%" PRIx64, mNodeId);
+        CommissioneeDeviceProxy * device = CurrentCommissioner().FindCommissioneeDevice(mNodeId);
```
Comment (cecille) [inline examples/chip-tool/commands/clusters/ModelCommand.cpp:41]: This can return nullptr if you haven't opened a pase connection, so that should be checked before sending the command.

File: examples/chip-tool/commands/pairing/PairingCommand.cpp [hunk 2]
```diff
@@ -89,7 +89,15 @@ CommissioningParameters PairingCommand::GetCommissioningParameters()
 
 CHIP_ERROR PairingCommand::PaseWithCode(NodeId remoteId)
 {
-    return CurrentCommissioner().EstablishPASEConnection(remoteId, mOnboardingPayload);
+    chip::SetupPayload payload;
+
+    QRCodeSetupPayloadParser(mOnboardingPayload).populatePayload(payload);
```
Comment (cecille) [inline examples/chip-tool/commands/pairing/PairingCommand.cpp:94]: Why was this changed to force BLE?

File: examples/chip-tool/README.md [hunk 3]
```diff
@@ -579,3 +579,25 @@ To learn more about the tool, how to build it, use its commands and advanced
 features, read the following guide:
 
 -   [Working with the CHIP Tool](https://github.com/project-chip/connectedhomeip/tree/master/docs/guides/chip_tool_guide.md)
+
+## Using the Client via PASE
+
+For testing chip-tool can stop after setting up a PASE session. Model commands
+can be issues via PASE as well.
```
Comment (bzbarsky-apple) [inline examples/chip-tool/README.md:586]:
```suggestion
can be issued via PASE as well.
```

File: examples/chip-tool/README.md [hunk 4]
```diff
@@ -579,3 +579,25 @@ To learn more about the tool, how to build it, use its commands and advanced
...
+
+```bash
+# Start chip-tool interactively
+./out/debug/chip-tool interactive start
+
+# Pair PASE only
+pairing qrcode-paseonly 7676777 MT:6FCJ142C00X50648G00
```
Comment (bzbarsky-apple) [inline examples/chip-tool/README.md:593]: Please use things like `${NODE_ID}` (of that's what the number there is?) and `${QR_CODE}` instead of hardcoding values that don't make sense in all settings.

File: examples/chip-tool/README.md [hunk 5]
```diff
@@ -579,3 +579,25 @@ To learn more about the tool, how to build it, use its commands and advanced
...
+
+# Issue a command via PASE
+onoff toggle 7676777 1 --via-pase 1
+
+# Also uses PASE
+onoff toggle 7676777 1
+
+# Switch back to CASE (fails if paseonly)
+onoff toggle 7676777 1 --via-pase 1
```
Comment (bzbarsky-apple) [inline examples/chip-tool/README.md:602]: Why is this switching back to CASE?

File: examples/chip-tool/README.md [hunk 6]
```diff
@@ -579,3 +579,25 @@ To learn more about the tool, how to build it, use its commands and advanced
...
+./out/debug/chip-tool interactive start
+
+# Pair PASE only
+pairing qrcode-paseonly 7676777 MT:6FCJ142C00X50648G00
+
+# Issue a command via PASE
+onoff toggle 7676777 1 --via-pase 1
+
+# Also uses PASE
+onoff toggle 7676777 1
```
Comment (bzbarsky-apple) [inline examples/chip-tool/README.md:599]: Why does this also use PASE?

File: examples/chip-tool/commands/pairing/PairingCommand.cpp [hunk 7]
```diff
@@ -89,7 +89,15 @@ CommissioningParameters PairingCommand::GetCommissioningParameters()
...
+    mSetupPINCode  = payload.setUpPINCode;
+    mDiscriminator = payload.discriminator;
+
+    RendezvousParameters rendezvousParams =
+        RendezvousParameters().SetSetupPINCode(mSetupPINCode).SetDiscriminator(mDiscriminator).SetPeerAddress(PeerAddress::BLE());
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/pairing/PairingCommand.cpp:99]: Won't this break PASE with code to onnetwork things?  Why are the changes here needed?

File: src/controller/CHIPDeviceController.h [hunk 8]
```diff
@@ -712,6 +712,7 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
 
     void HandleAttestationResult(CHIP_ERROR err);
 
+public:
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.h:715]: Why?  There's a public API for this (`GetDeviceBeingCommissioned`), which incidentally makes you think about the failure case @cecille raised because it returns an error in that case.  There's a reason these methods are not public.

File: examples/chip-tool/README.md [hunk 9]
```diff
@@ -579,3 +579,25 @@ To learn more about the tool, how to build it, use its commands and advanced
 features, read the following guide:
 
 -   [Working with the CHIP Tool](https://github.com/project-chip/connectedhomeip/tree/master/docs/guides/chip_tool_guide.md)
+
+## Using the Client via PASE
```
Comment (tcarmelveilleux) [inline examples/chip-tool/README.md:583]: PASE is only for commissioning. While the commissioning protocol makes use of commands over PASE, and indeed, anything can be done against the device during commissioning, you cannot rely on PASE forever: on PASE establishment, a fail-safe timer will be set, which will cause eventual reset of state. Also, the commissioning window can only be open for 15 minutes at a time. Could you please explain why this new change is encouraging *operational* use of PASE over BLE for non-commissioning use-cases? Illustrating this without the proper warnings may cause SDK users to believe they can use PASE over BLE "forever" to control devices operationally, which is explicitly a problem by spec (due to aforementioned commissioning window behavior).

**Review by tcarmelveilleux — CHANGES_REQUESTED** [review]
- [ ] Cannot encourage operational use of PASE. The caveats of PASE-only chip-tool usage must be mentioned. It's OK to have such a mode to use chip-tool to do some additional/different configs rather than just the "default" commissioning logic, but we must be careful not to imply that a product could forever just be controllable over BLE.

File: examples/chip-tool/README.md (hunk 9 above)
Comment (cecille) [reply examples/chip-tool/README.md:583]: I spoke with Marcus about this on slack - they would like to use the pase connection to get ID information about the device before commissioning. This is appropriate for such a use case. Adding this in the documentation would be beneficial.

File: examples/chip-tool/README.md (hunk 6 above)
Comment (markus-becker-tridonic-com) [reply examples/chip-tool/README.md:599]: The --via-pase sets the member variable mViaPase which keeps the value.

File: examples/chip-tool/README.md (hunk 5 above)
Comment (markus-becker-tridonic-com) [reply examples/chip-tool/README.md:602]: Should have been
`onoff toggle 7676777 1 --via-pase 0`

Will correct.

File: examples/chip-tool/README.md (hunk 4 above)
Comment (markus-becker-tridonic-com) [reply examples/chip-tool/README.md:593]: Will change this in an upcoming commit.

File: examples/chip-tool/README.md (hunk 9 above)
Comment (markus-becker-tridonic-com) [reply examples/chip-tool/README.md:583]: @tcarmelveilleux 

it's understood that PASE is not to be used for true operational use-cases. however, i see several uses:
* usage of PASE for end-of-line production tests with manufacturer specific clusters
* usage of PASE for identifying the device while commissioning

i will add the disclaimers to the README.

File: examples/chip-tool/commands/pairing/PairingCommand.cpp (hunk 2 above)
Comment (markus-becker-tridonic-com) [reply examples/chip-tool/commands/pairing/PairingCommand.cpp:94]: Undone in latest commits.

File: examples/chip-tool/commands/clusters/ModelCommand.cpp (hunk 1 above)
Comment (markus-becker-tridonic-com) [reply examples/chip-tool/commands/clusters/ModelCommand.cpp]: Checked now and using the public API.

File: src/controller/CHIPDeviceController.h (hunk 8 above)
Comment (markus-becker-tridonic-com) [reply src/controller/CHIPDeviceController.h]: Using the public API now.

File: examples/chip-tool/commands/pairing/PairingCommand.cpp (hunk 7 above)
Comment (markus-becker-tridonic-com) [reply examples/chip-tool/commands/pairing/PairingCommand.cpp]: Removed the SetPeerAddress(). Is that enough @bzbarsky-apple ?

File: examples/chip-tool/commands/clusters/ModelCommand.cpp [hunk 10]
```diff
@@ -35,11 +35,22 @@ CHIP_ERROR ModelCommand::RunCommand()
...
+
+        return this->SendCommand(commissioneeDevice, this->mEndPointId);
+    }
+
     ChipLogProgress(chipTool, "Sending command to node 0x%" PRIx64, mNodeId);
     return CurrentCommissioner().GetConnectedDevice(mNodeId, &mOnDeviceConnectedCallback, &mOnDeviceConnectionFailureCallback);
 }
 
-void ModelCommand::OnDeviceConnectedFn(void * context, ChipDevice * device)
+void ModelCommand::OnDeviceConnectedFn(void * context, chip::OperationalDeviceProxy * device)
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/ModelCommand.cpp:53]:
```suggestion
void ModelCommand::OnDeviceConnectedFn(void * context, OperationalDeviceProxy * device)
```

File: examples/chip-tool/commands/pairing/PairingCommand.cpp [hunk 11]
```diff
@@ -89,7 +89,14 @@ CommissioningParameters PairingCommand::GetCommissioningParameters()
...
 {
-    return CurrentCommissioner().EstablishPASEConnection(remoteId, mOnboardingPayload);
+    chip::SetupPayload payload;
+
+    QRCodeSetupPayloadParser(mOnboardingPayload).populatePayload(payload);
+    mSetupPINCode  = payload.setUpPINCode;
+    mDiscriminator = payload.discriminator;
+
+    RendezvousParameters rendezvousParams = RendezvousParameters().SetSetupPINCode(mSetupPINCode).SetDiscriminator(mDiscriminator);
+    return CurrentCommissioner().EstablishPASEConnection(remoteId, rendezvousParams);
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/pairing/PairingCommand.cpp:99]: This still looks like it will break things.  The point is that the version of `EstablishPASEConnection` that takes params assumes that discovery has already happened and the params have the right address to use.

Why are we changing this function at all?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/chip-tool/commands/pairing/PairingCommand.cpp (hunk 7 above)
Comment (bzbarsky-apple) [reply examples/chip-tool/commands/pairing/PairingCommand.cpp]: Did this get tested?  Why ask me if it's enough?
